### PR TITLE
Stop using sklearn's private `_scorer` API for custom metrics in SKLL. 

### LINKED ISSUE #748: Update scikit-learn dependency to v1.3.0
v1.3.0 is [out](https://scikit-learn.org/stable/whats_new/v1.3.html#changes-1-3). We should update SKLL to use it since we are doing a new release anyway.

### LINKED ISSUE #750: Overhaul how custom metrics are used in SKLL
Starting with scikit-learn v1.3.0, it's no longer possible to simply update the `sklearn.metrics._scorer._SCORERS` dictionary with new metric functions and have them be picked up. This is because scikit-learn now implements additional parameter validation and rejects any metric names that aren't its own metrics. To use custom metrics, we basically have to use callables instead of relying on custom strings that get looked up in `_SCORERS`. This means that we have to overhaul how we use custom metrics in SKLL (not just user-defined custom metrics but also pre-defined custom metrics such as `pearson` and `kappa` etc.). 

This blocks #748.

### PR DESCRIPTION
**Pull request by desilinguist**
- Update scikit-learn to v1.3.0.
- Move all custom pre-defined metrics to `_PREDEFINED_CUSTOM_METRICS` dictionary in `metrics.py`.
- Make a copy of the above dictionary as `CUSTOM _METRICS`. This dictionary will be the source of all custom metrics: pre-defined and user-defined.
- We need two dictionaries because at some points in the code, we need to be able to differentiate between pre-defined custom metrics and user-defined custom metrics.
- Remove any use of `sklearn.metrics._scorer` private API and use `sklearn.metrics.get_scorer()` and `sklearn.metrics.get_scorer_names()` instead.
- Use the actual metric function when using custom metrics instead of its name. This is the core change that makes everything work since sklearn validates callables but does not validate custom metric name strings.
- Update tests to not use `_SCORERS` and fix a minor bug in another test related to scikit-learn v1.3.0. 

This PR closes #748 and closes #750. 

To review this PR, please try to create some [custom metrics](https://skll.readthedocs.io/en/latest/custom_metrics.html) for both the titanic and California examples and try to use them via both the API and the configuration file.

### CODE REVIEW

**Review by Frost45 — COMMENTED** [review]
I would like to suggest a couple of changes to the custom_metrics documentation:

1. Fix indentation in [custom.py](https://skll.readthedocs.io/en/latest/custom_metrics.html#id1). There are a few spaces at the start of every line.
2. Change `log` to `logs` in the config file on the same page. Otherwise, you get a KeyError.

**Review by Frost45 — APPROVED** [review]
Looks great! Tested it out with multiple custom metrics.

File: skll/metrics.py [hunk 1]
```diff
@@ -270,11 +273,11 @@ def register_custom_metric(custom_metric_path: PathOrStr, custom_metric_name: st
 
     # once we know that the module name is okay, we need to make sure
     # that the metric function name is also okay
-    if custom_metric_name in _SCORERS:
+    if custom_metric_name in get_scorer_names() or custom_metric_name in _CUSTOM_METRICS:
```
[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/metrics.py:276]: Yeah, this would only happen if `register_custom_metric()` gets called and that will only happen in an experiment if (from `experiments.py`):

```
            if (
                metric_name not in get_scorer_names()
                and metric_name not in _CUSTOM_METRICS
                and metric_name is not None
            ):
```

So, overriding a scikit-learn metric name in a config file won't trigger this. However, if I use the same `custom.py` as in the docs and rename `f075` to `f1_score_micro`, then calling `register_custom_metric()` will raise this error.

[the remark this answers is not in the document]
Comment (desilinguist) [reply skll/metrics.py:276]: May be it's worth adding a warning the docs that overriding scikit-learn metrics will not work whether using experiments or the API.

**Comment by mulhod** [comment]
Tests passed for me just now. Will review shortly.
